lib/http.c

```diff
@@ -1662,6 +1662,50 @@ enum proxy_use {
   HEADER_CONNECT  /* sending CONNECT to a proxy */
 };
 
+/* used to compile the provided trailers into one buffer */
+CURLcode Curl_http_compile_trailers(struct connectdata *conn,
+                                    struct curl_slist *trailers,
+                                    Curl_send_buffer *buffer)
+{
+  char *ptr = NULL;
+  CURLcode result = CURLE_OK;
+  struct Curl_easy *data = conn->data;
+  const char *endofline_native = NULL;
+  const char *endofline_network = NULL;
+
+  /* basically, loop through trailers, make sure they all are in 'X: Y' format
+     and then push them into the buffer. */
+  /* TODO: Maybe divide Curl_add_custom_headers to make it reusable here */
+
+  /* The way we send trailing headers is no different from the way we send
+     normal data. So we have to check for this, as it can break the trailers
+     if the conversion happens later.*/
+
+  if(data->set.crlf) {
+    /* \n will become \r\n later on */
+    endofline_native  = "\n";
+    endofline_network = "\x0a";
+  }
+  else {
+    endofline_native  = "\r\n";
+    endofline_network = "\x0d\x0a";
+  }
+
+  while(trailers) {
+    /* only add correctly formatted trailers */
+    ptr = strchr(trailers->data, ':');
+    if(ptr) {
+      result = Curl_add_bufferf(buffer, "%s%s", trailers->data,
+              endofline_native);
+      if(result)
+        return result;
+    }
+    trailers = trailers->next;
+  }
+  Curl_add_buffer(buffer, endofline_network, 2);
+  return result;
```

Comment:
Should it also free the list here before returning?

Reply:
I think I mentioned it in the docs, if not I will add it. The string could be allocated with some mechanism other than plain malloc, so we could either force the user to pass something upon which we could use free, or to just memcpy it (as is done in Curl_add_buffer if I recall correctly), and leave it to the user to free the list after the transfer has ended. _(I may be wrong in my reasoning)_

I still have a problem since I used `Curl_send_buffer` to avoid rewriting that code, but that is linked to the HTTP protocol _(which is why the --disable-http test fails)_, so I think I'll have to work around that.

Reply:
The callback can't and won't be invoked if there's no HTTP support. You can just use a strategic `#ifndef CURL_DISABLE_HTTP` on the right place.

Reply:
I'm really sorry, I completely forgot about that ... Will fix in the next commit.